pkg/controllers/scheduling/enqueue.go

```diff
@@ -0,0 +1,216 @@
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	queue.Add(key)
+}
+
+func (e *enqueuer) enqueueClusterSetBinding(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	namespace, _, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	// enqueue all placement that ref to the binding
+	objs, err := e.placementIndexer.ByIndex(placementsByClusterSetBinding, key)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	anyObjs, err := e.placementIndexer.ByIndex(placementsByClusterSetBinding, fmt.Sprintf("%s/%s", namespace, anyClusterSet))
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	objs = append(objs, anyObjs...)
+
+	for _, o := range objs {
+		placement := o.(*clusterapiv1beta1.Placement)
+		klog.V(4).Infof("enqueue placement %s/%s, because of binding %s", placement.Namespace, placement.Name, key)
+		e.enqueuePlacementFunc(placement, e.queue)
+	}
+}
+
+func (e *enqueuer) enqueueClusterSet(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	objs, err := e.clusterSetBindingIndexer.ByIndex(clustersetBindingsByClusterSet, key)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	for _, o := range objs {
+		clusterSetBinding := o.(*clusterapiv1beta2.ManagedClusterSetBinding)
+		klog.V(4).Infof("enqueue clustersetbinding %s/%s, because of clusterset %s", clusterSetBinding.Namespace, clusterSetBinding.Name, key)
+		e.enqueueClusterSetBinding(clusterSetBinding)
+	}
+}
+
+func (e *enqueuer) enqueueCluster(obj interface{}) {
+	cluster, ok := obj.(*clusterapiv1.ManagedCluster)
+	if !ok {
+		runtime.HandleError(fmt.Errorf("obj %T is not a ManagedCluster", obj))
+		return
+	}
+
+	clusterSets, err := clusterapiv1beta2.GetClusterSetsOfCluster(cluster, e.clusterSetLister)
+	if err != nil {
+		klog.V(4).Infof("Unable to get clusterSets of cluster %q: %v", cluster.GetName(), err)
+		return
+	}
+
+	for _, clusterSet := range clusterSets {
+		klog.V(4).Infof("enqueue clusterSet %s, because of cluster %s", clusterSet.Name, cluster.Name)
+		e.enqueueClusterSet(clusterSet)
+	}
+}
+
+func (e *enqueuer) enqueuePlacementScore(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	objs, err := e.placementIndexer.ByIndex(placementsByScore, name)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	for _, o := range objs {
+		placement := o.(*clusterapiv1beta1.Placement)
+		klog.V(4).Infof("enqueue placement %s/%s, because of score %s", placement.Namespace, placement.Name, key)
+		e.enqueuePlacementFunc(placement, e.queue)
```

Comment:
Should we consider the access scope of the placement? For example, if the score belongs to a cluster on which this placement has no permission, we can just skip this placement.